model_utilities/train/finetuner.py

```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         plt.legend()
         plt.savefig("./debug/graphs/losses.png")
 
-    def __train(self, model: AutoModelForSeq2SeqLM, tokenizer: NllbTokenizer, dataset: ds.Dataset, optimizer: Adafactor,
+    def __tokenize_dataset(self, dataset: ds.Dataset, tokenizer: NllbTokenizerFast, max_length: int) -> Dataset:
+        # Not meant to be called outside of this function
+        def tokenize_fn(dataset: ds.Dataset, src_lang: str, tgt_lang: str, tokenizer: NllbTokenizerFast, max_length: int) -> BatchEncoding:
+            tokenizer.src_lang = src_lang
+            tokenizer.tgt_lang = tgt_lang
+            tokenized_input = tokenizer(dataset[src_lang], text_target=dataset[tgt_lang], return_tensors='pt', padding=True, truncation=True, max_length=max_length)
+
+            # -100 is the default value ignored by pytorch
+            # TODO: Check performance after removing this line
+            tokenized_input.labels[tokenized_input.labels == tokenizer.pad_token_id] = -100
+            return tokenized_input
+
+        tokenized_train_dataset_pl_to_csb = dataset.map(
+            lambda row: tokenize_fn(row, "pol_Latn", "csb_Latn", tokenizer, max_length),
+            batched=True
+        )
+        tokenized_train_dataset_csb_to_pl = dataset.map(
+            lambda row: tokenize_fn(row, "csb_Latn", "pol_Latn", tokenizer, max_length),
+            batched=True
+        )
+
+        tokenized_dataset = ds.interleave_datasets([tokenized_train_dataset_pl_to_csb, tokenized_train_dataset_csb_to_pl])
```

Comment:
Why do we interleave the datatests?

Reply:
The previous approach was to tokenize after we randomize the direction in which we are translating (literally pick a pair and then pass one as input and one as labels to the model at random)

Because we moved tokenization outside the training loop we thought it would be better to just train the model in both directions for each sentence. It was easier to merge the duplicated dataset to preserve the inputs being in the input_ids column and labels in the labels column than to swap it around during the training loop.

We could concatenate the datasets instead of interleaving them but I thought it's just nicer this way and maybe more closely represents the previous implementation.

No idea if it will yield better results than concat though. We could try both?